### PR TITLE
Add tests of various features

### PR DESCRIPTION
**Pull request by TorkelE**
Ok, I have tried to rewrite various tests in Catalyst to work within MTK (a bit more complicated since each system have to be generated directly, and cannot be generated from a single `ReactionSystem` model. These tests are primarily going through vaiours ways to interact with SciML structures, and giving input to problems, across various system types.

I have tried to go through all broken tests, and mark issues that are related. Unfortunately there are quite a few, so someone else would probably need to have a second look at each to check the issues are accurate.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
The tests fail?

**Comment by TorkelE** [comment]
Not too surprised. I will have a look locally (the test log is too large to actually see what is going on)

**Comment by TorkelE** [comment]
Ok, tests pass now (except for one downstream test)

**Review by AayushSabharwal — COMMENTED** [review]
Thanks for the extremely comprehensive tests. Apart from the specific comments below, I have a couple of overarching requests.

Firstly, these tests don't belong in MTK. They are all testing features implemented in SciMLBase, and should go there alongside the other tests for similar functionality.

- Anything to do with `remake` should be in SciMLBase's `modelingtoolkit_remake.jl` testset
- All of the tests for symbolic indexing/interpolation of integrators/solutions/etc should be in a `comprehensive_indexing.jl` testset
- Both of the above should be in the `"SymbolicIndexingInterface"` group

That way all of these will be run in MTK downstream, and are tested in the library that implements the features being tested. SciMLBase wants to run these tests because changes in SciMLBase are more likely to break them than ones in MTK. `remake` for example is almost entirely a SciMLBase feature, and the only places MTK is likely to break it is `remake_buffer`.

Testing `save_idxs` is also unnecessary in my opinion. It's not something that's broken, it's a feature we don't support yet.

I would also like to not merge these tests in a state where the broken ones are commented out/marked as `@test_broken`. That defeats the purpose of the tests. We should test everything, and merge when it's all fixed.

File: test/sciml_struct_interfacing.jl [hunk 1]
```diff
@@ -0,0 +1,444 @@
+### Prepares Tests ###
+
+# Fetch packages
+using ModelingToolkit, JumpProcesses, NonlinearSolve, OrdinaryDiffEq, Plots,
+      SteadyStateDiffEq, StochasticDiffEq, Test
+using ModelingToolkit: t_nounits as t, D_nounits as D
+using ModelingToolkit: getp, getu, setp, setu
```
Comment (AayushSabharwal) [inline test/sciml_struct_interfacing.jl:7]: Also just noticed this. These should be imported from SII, not MTK. They're not part of MTK's public API.

**Comment by AayushSabharwal** [comment]
I should have mentioned: I'll move the tests when I PR to SciMLBase to fix the bugs

File: test/sciml_problem_inputs.jl [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+# Perform SDE simulations (singular and ensemble).
+let
+    # Creates normal and ensemble problems.
+    base_sprob = SDEProblem(ssys, u0_alts[1], tspan, p_alts[1])
+    base_sol = solve(base_sprob, ImplicitEM(); seed, saveat = 1.0)
+    base_eprob = EnsembleProblem(base_sprob)
+    base_esol = solve(base_eprob, ImplicitEM(); seed, trajectories = 2, saveat = 1.0)
+
+    # Simulates problems for all input types, checking that identical solutions are found.
+    @test_broken false # first remake in subsequent test yields a `ERROR: type Nothing has no field portion`.
```
[the remark this answers is not in the document]
Comment (TorkelE) [reply test/sciml_problem_inputs.jl:118]: I think this ones goes through a couple of additional cases (e.g. observables). not sure why this one errors though, probably something very niche.

File: test/sciml_problem_inputs.jl [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+let
+    # Creates normal and ensemble problems.
+    base_dprob = DiscreteProblem(jsys, u0_alts[1], tspan, p_alts[1])
+    base_jprob = JumpProblem(jsys, base_dprob, Direct(); rng)
+    base_sol = solve(base_jprob, SSAStepper(); seed, saveat = 1.0)
+    base_eprob = EnsembleProblem(base_jprob)
+    base_esol = solve(base_eprob, SSAStepper(); seed, trajectories = 2, saveat = 1.0)
+
+    # Simulates problems for all input types, checking that identical solutions are found.
+    @test_broken false # first remake in subsequent test yields a `ERROR: type Nothing has no field portion`.
```
[the remark this answers is not in the document]
Comment (TorkelE) [reply test/sciml_problem_inputs.jl:137]: This on fails, but yes, let's investigate closer.

File: test/sciml_problem_inputs.jl [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+        (k1 => 1.0, k2 => 2.0, k3 => 3.0)
+    ]
+
+    # Loops through all potential parameter sets, checking their inputs yield errors.
+    # Broken tests are due to this issue: https://github.com/SciML/ModelingToolkit.jl/issues/2779
+    for ps in [ps_valid; ps_invalid], u0 in [u0_valid; u0s_invalid]
+        # Handles problems with/without tspan separately. Special check ensuring that valid inputs passes.
+        for (xsys, XProblem) in zip(
+            [osys, ssys, jsys], [ODEProblem, SDEProblem, DiscreteProblem])
+            if (ps == ps_valid) && (u0 == u0_valid)
```
[the remark this answers is not in the document]
Comment (TorkelE) [reply test/sciml_problem_inputs.jl]: Thanks for pointing this out (technically no in the updated version, but you are right that there is a bug here that I missed!)

File: test/sciml_struct_interfacing.jl [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+        rp = remake(prob; p = [sys.k2 => 14.0])
+        @test rp.ps[[kp, kd, k1, k2]] == [1.0, 0.1, 0.25, 14.0]
+        rp = remake(prob; p = [:k2 => 15.0])
+        @test rp.ps[[kp, kd, k1, k2]] == [1.0, 0.1, 0.25, 15.0]
+    end
+end
+
+# Test integrator indexing.
+let
+    @test_broken false # NOTE: Multiple problems for `nint` (https://github.com/SciML/SciMLBase.jl/issues/662).
```
[the remark this answers is not in the document]
Comment (TorkelE) [reply test/sciml_struct_interfacing.jl:195]: The issues got closed but then the conversation continued. But I see that there was a PR merged related. I checked fairly recently and this was still broken, but should check again.

File: test/sciml_struct_interfacing.jl (hunk 1 above)
Comment (TorkelE) [reply test/sciml_struct_interfacing.jl]: Good point, I've updated this

**Comment by TorkelE** [comment]
> I should have mentioned: I'll move the tests when I PR to SciMLBase to fix the bugs

Thanks!

**Comment by TorkelE** [comment]
Yeah, I am a bit uncertain what actually goes where now when the MTK pipeline have been split up in so many different places. To me it is basically tests of the System types that MTK implements. But I agree that if possible it would make sense to move them as far internal as possible. If you point out what and where I could do it, but might be easier for you to do it yourself.

**Comment by TorkelE** [comment]
For the `save_idx` I think it is just that it has been broken since the very beginning. It is something that should be supported, so don't think it is bad to keep the tests in until it actually gets fixed, even if it might be a while.

File: test/sciml_struct_interfacing.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (TorkelE) [reply test/sciml_struct_interfacing.jl:195]: Checked, this still fails for nonlinear integrators

**Comment by TorkelE** [comment]
Are we sure that the modified version of these that got ported to SciMLBase are actually exhaustive? I.e. regression that I caught with these tests in Catalyst still appear in the ecosystem: https://github.com/SciML/ModelingToolkit.jl/issues/2838

**Comment by TorkelE** [comment]
I have modified this one, removing the SciMLBase related stuff, and also adding tests that static vectors can be used as input across all problem types. @AayushSabharwal

**Comment by TorkelE** [comment]
Haven't heard anything in a while, can we merge this @AayushSabharwal @ChrisRackauckas ?

File: test/sciml_problem_inputs.jl [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+    base_eprob = EnsembleProblem(base_jprob)
+    base_esol = solve(base_eprob, SSAStepper(); seed, trajectories = 2, saveat = 1.0)
+
+    # Simulates problems for all input types, checking that identical solutions are found.
+    @test_broken false # first remake in subsequent test yields a `ERROR: type Nothing has no field portion`.
+    for u0 in u0_alts, p in p_alts
+        #    jprob = remake(base_jprob; u0, p)
+        #    @test base_sol == solve(base_jprob, SSAStepper(); seed, saveat = 1.0)
+        #    eprob = remake(base_eprob; u0, p)
+        #    @test base_esol == solve(eprob, SSAStepper(); seed, trajectories = 2, saveat = 1.0)
```
Comment (ChrisRackauckas) [inline test/sciml_problem_inputs.jl:154]: this isn't actually ran

File: test/sciml_problem_inputs.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChrisRackauckas) [reply test/sciml_problem_inputs.jl:130]: this test broken isn't setup right

**Comment by ChrisRackauckas** [comment]
This needs a bit of work. All of the `@test_broken`s are not setup correctly and the actual tests are commented.

**Comment by TorkelE** [comment]
So, there are a large number of cases where I do
```julia
    @test_broken false # Does not work for certain inputs, likely related to https://github.com/SciML/ModelingToolkit.jl/issues/2804.
    for u0 in u0_alts_vec, p in p_alts_vec
        oprob = remake(base_oprob; u0, p)
        # @test base_sol == solve(oprob, Tsit5(); saveat = 1.0)
        eprob = remake(base_eprob; u0, p)
        # @test base_esol == solve(eprob, Tsit5(); trajectories = 2, saveat = 1.0)
    end
```
because some of the tests in the loop are broken, and some are not. Then just doing either `@test` or `@test_broken` will cause test failure. Separating the cases will cause _tons_ of line to be written. Hence I comment out the actual test lines, and then write 
```julia
@test_broken false
```
to record in the test file that there is something going on, so that the case where something is broken is lost. Similar in cases where the full test is commented out (as in some cases something in the build up fails, and sometimes not).

**Comment by ChrisRackauckas** [comment]
Then why is it important for this to merge if the test broken isn't actually set to tell us if they are still broken? That's the only point of `@test_broken`.

**Comment by TorkelE** [comment]
It is meant as a reminder when you run the tests, it tells you that there are some broken tests and where (and links the issue).

Do you want me to just remove the `@test_broken` and comment out the relevant bits?

**Comment by TorkelE** [comment]
@ChrisRackauckas I removed the stuff with the broken test, we can add them back later when they work. This should be good now. @AayushSabharwal do you have any last comments?

**Comment by ChrisRackauckas** [comment]
Using `@test_broken` is fine, but the way you did it would not give an appropriate error when the test is fixed, and so it wasn't actually useful. It would be good to add correct `@test_broken`s where appropriate.

**Pull request merged**